src/lightning/data/cache/cache.py

```diff
@@ -0,0 +1,98 @@
+# Copyright The Lightning AI team.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+from typing import Any, Dict, Optional
+
+import torch
+
+from lightning.data.cache.pytree import tree_flatten
+from lightning.data.cache.reader import BinaryReader
+from lightning.data.cache.writer import BinaryWriter
+from lightning.data.datasets.env import _DistributedEnv
+
+logger = logging.Logger(__name__)
+
+
+class Cache:
+    def __init__(
+        self,
+        cache_dir: str,
+        compression: Optional[str] = None,
+        chunk_size: int = 2 << 26,
+    ):
+        """The Cache enables to optimise dataset format for cloud training. This is done by grouping several elements
+        together in order to accelerate fetching.
+
+        Arguments:
+            cache_dir: The path to where the chunks will be stored.
+            compression: The name of the algorithm to reduce the size of the chunks
+            chunk_size: The maximum byte size of chunk.
+
+        """
+        super().__init__()
+        self._writer = BinaryWriter(cache_dir, chunk_size=chunk_size, compression=compression)
+        self._reader = BinaryReader(cache_dir, compression=compression)
+        self._cache_dir = cache_dir
+        self._is_done = False
+        self._distributed_env = _DistributedEnv.detect()
+        self._num_workers: Optional[int] = None
+
+    def _is_item_equal(self, data_1: Any, data_2: Any) -> bool:
+        data_1_flattened, _ = tree_flatten(data_1)
+        data_2_flattened, _ = tree_flatten(data_2)
+
+        if len(data_1_flattened) != len(data_2_flattened):
+            return False
+
+        return all(self._is_data_equal(d1, d2) for d1, d2 in zip(data_1_flattened, data_2_flattened))
+
+    def _is_data_equal(self, d1, d2) -> bool:
+        if isinstance(d1, torch.Tensor) and isinstance(d2, torch.Tensor):
+            return torch.equal(d1, d2)
+        return d1 == d2
```

Comment:
why can't you use d1 == d2 for tensors? It should literally call `torch.equal`. Probably something like this makes more sense:

```suggestion
        equality = (d1 == d2)
        if isinstance(equality, torch.Tensor):
            return equality.all()
        return equality
```

Reply:
Done.